### ISSUE TITLE
LayoutReader: How can I inference if I only have layout information data?

### ISSUE THREAD

**Issue by WeihongM** [microsoft/unilm#559]
Hello, thanks for the impressive work.
When I do decode_seq2seq, it seems I need to prepare the json file with text information. I want to ask how I can do inference process when only layout information is provided. (I need to change the feature extracting rather than use function "load_and_cache_layoutlm_examples")
Which code I need to modify. Hope you can give me some suggestions.

Thanks.
@HYPJUDY

**Comment by HYPJUDY** [comment]
Hi, maybe you are asking about the `layoutlm_only_layout` setting? Please check [this issue and reply](https://github.com/microsoft/unilm/issues/464#issuecomment-946272607).

**Comment by WeihongM** [comment]
@HYPJUDY  Hello, thanks for your kindly reply.
I mean if there is only layout information is provided, how can I finish the inference process? Sorry, I don't have text information. Therefore, it seems I don't need to calculate BLEU score too.
[This line](https://github.com/microsoft/unilm/blob/master/layoutreader/decode_seq2seq.py#L232) in the code will run error. 
Can I use this [function](https://github.com/microsoft/unilm/blob/master/layoutreader/s2s_ft/utils.py#L321) directly? Which code should I modify?

Hope for your reply.
Thanks!

**Comment by myh12138** [comment]
> @HYPJUDY Hello, thanks for your kindly reply. I mean if there is only layout information is provided, how can I finish the inference process? Sorry, I don't have text information. Therefore, it seems I don't need to calculate BLEU score too. When I run the inference process with layout information only, [This line](https://github.com/microsoft/unilm/blob/master/layoutreader/decode_seq2seq.py#L232) in the code will run error. Can I use this [function](https://github.com/microsoft/unilm/blob/master/layoutreader/s2s_ft/utils.py#L321) directly? Which code should I modify?
> 
> Hope for your reply. Thanks!

Are you success?
when I change input_folder to my data and make layoutlm_only_layout ,the progress still shows the original data?
Can you give me some suggestions? Thanks

**Comment by zlwang-cs** [comment]
@WeihongM @myh12138  Hi, thanks for your interests in our work! In our problem settings, we have both text and layout. The `layoutlm_only_layout` is for ablation study only, which means we have text but just do not encode it in the model. In your case (no text at all), I would suggest you to generate some random words as placeholders (for example 1 2 3 4 ...). I think it should be the easiest way to run the code (but I am not sure whether there will be some other issues). I would also suggest you to modify a little bit on the evaluation codes to make sure the results are reasonable.

**Comment by WeihongM** [comment]
OK, thanks for your reply.

**Comment by WeihongM** [comment]
Hello, I think shuffling the sentence is more general as input which means I don't prepare the input order strictly.
However, I find the loss is difficult to decrease when the sentence_shuffle_rate is set to 1.0.  Do you have other code changes when shuffling the sentence?
@HYPJUDY @zlwang-cs

**Comment by zlwang-cs** [comment]
> Hello, I think shuffling the sentence is more general as input which means I don't prepare the input order strictly.
> However, I find the loss is difficult to decrease when the sentence_shuffle_rate is set to 1.0. Do you have other code changes when shuffling the sentence?
> @HYPJUDY @zlwang-cs

No, the `sentence shuffle rate` only affects the word order in the input sentences.

**Comment by WeihongM** [comment]
Thanks for your reply. The loss is stable around at 4.5. It seems strange. I will try to figure out the problem.
In order to speed up the training process for the limitation of GPU number, I only use dataset-train-s2s-layout-m1.json for training. I am not sure whether the data size cause this difference?

**Comment by zlwang-cs** [comment]
> 

I haven't tried this setting but I guess the dataset size shouldn't be a problem. The m1 m2 m3... are divided randomly (we have several machines to process the data so we have several parts)

**Comment by myh12138** [comment]
> @WeihongM @myh12138 Hi, thanks for your interests in our work! In our problem settings, we have both text and layout. The `layoutlm_only_layout` is for ablation study only, which means we have text but just do not encode it in the model. In your case (no text at all), I would suggest you to generate some random words as placeholders (for example 1 2 3 4 ...). I think it should be the easiest way to run the code (but I am not sure whether there will be some other issues). I would also suggest you to modify a little bit on the evaluation codes to make sure the results are reasonable.

Thanks for your reply,I will try it later. I also want to know what does "src" mean and "tgt" mean, if I try to test some images how should I make "src" and "tgt".

**Comment by zlwang-cs** [comment]
> > @WeihongM @myh12138 Hi, thanks for your interests in our work! In our problem settings, we have both text and layout. The `layoutlm_only_layout` is for ablation study only, which means we have text but just do not encode it in the model. In your case (no text at all), I would suggest you to generate some random words as placeholders (for example 1 2 3 4 ...). I think it should be the easiest way to run the code (but I am not sure whether there will be some other issues). I would also suggest you to modify a little bit on the evaluation codes to make sure the results are reasonable.
> 
> Thanks for your reply,I will try it later. I also want to know what does "src" mean and "tgt" mean, if I try to test some images how should I make "src" and "tgt".

`src` and `tgt` mean the source and target in sequence-to-sequence model

**Comment by myh12138** [comment]
> 

Thanks.

**Comment by WeihongM** [comment]
I re-create the layout only dataset. And training the program using this function [load_and_cache_line_order_examples](https://github.com/microsoft/unilm/blob/master/layoutreader/s2s_ft/utils.py#L321).
I found using layout and text with shuffle rate can converage, however, when only using the layout information with shuffling option, the loss is still high.
Can you help me figure out this problem? or if there anything wrong in this function [load_and_cache_line_order_examples](https://github.com/microsoft/unilm/blob/master/layoutreader/s2s_ft/utils.py#L321).
@zlwang-cs @HYPJUDY

**Comment by zlwang-cs** [comment]
Please try to use the `load_and_cache_layoutlm_examples` instead or make some your own modification.

**Comment by WeihongM** [comment]
Thanks for your reply.
1. In the paper, how do you experiment with the layout-only setting? 
2. What's more, how can I use this model when I only have dataset with layout?  Do I need to set  flag "layout_only_dataset" True, too?
Can I prepare the dataset with  some random words as placeholders (for example 1 2 3 4 ...) which mentioned above. The text information will not be used when use the option "layoutlm_only_layout", right?

**Comment by zlwang-cs** [comment]
1. When we do the layout-only setting, we only use the `layoutlm_only_layout` flag. We do not use the `layout_only_dataset` flag at all. (see https://github.com/microsoft/unilm/blob/b94ec76c36f02fb2b0bf0dcb0b8554a2185173cd/layoutreader/s2s_ft/modeling.py#L203)
2. Using the placeholders is my intuitive idea, which is not covered by our existing works. If you're interested, you definitely should have a try.

**Comment by WeihongM** [comment]
Thanks for your reply.
I found that the problem stems from the function: load_and_cache_line_order_examples. I have tried your advice and use the placeholders. The loss decrease normally.
Maybe the function "load_and_cache_line_order_examples" need to be checked.

**Comment by zlwang-cs** [comment]
Thanks for your comments. Glad to see your questions are solved.

**Comment by myh12138** [comment]
> Thanks for your comments. Glad to see your questions are solved.

while I trying to test my own dataset, I have change example_path(input_folder) to my dataset, however the result still from dataset-test-s2s-layout-m1.json

**Comment by JBBalling** [comment]
> while I trying to test my own dataset, I have change example_path(input_folder) to my dataset, however the result still from dataset-test-s2s-layout-m1.json

Hello there, 
I have the same problem, when changing the data-directory to some custom text and layout information, the model still predicts the output from the test_data, any suggestions on that?

@myh12138, @zlwang-cs

**Comment by zlwang-cs** [comment]
Could you give more details of your implementation, such as the arguments you are using?

**Comment by Yashaswi7** [comment]
Since the dataset and the object are now made custom, can you provide an example input image to run ocr on. I was using this but getting unidecode error.
```python decode_seq2seq.py --fp16 \
    --model_type layoutlm \
    --tokenizer_name bert-base-uncased \
    --input_folder /path/to/ReadingBank/test \
    --cached_feature_file /path/to/ReadingBank/features_test.pt \
    --output_file /path/to/output/LayoutReader/layoutlm/output.txt \
    --split test \
    --do_lower_case \
    --model_path /path/to/output/LayoutReader/layoutlm/ckpt-75000 \
    --max_seq_length 1024 \
    --max_tgt_length 511 \
    --batch_size 32 \
    --beam_size 1 \
    --length_penalty 0 \
    --forbid_duplicate_ngrams \
    --mode s2s \
    --forbid_ignore_word "."
```
I dowloaded the pretrained model from  [https://huggingface.co/nielsr/layoutreader-readingbank/tree/main](). Can you please confirm if this the right model or I need to format the input image in some particular format?

Thanks.
